_posts/2020-04-03-what-does-appsync-do.md

```diff
@@ -0,0 +1,200 @@
+    Properties:
+      ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+      Description: API key for graphQL Api
+```
+
+There are two resources in the template.
+
+Firstly, it is creating the resource for AppSyncAPI. There are few methods available for authentication type such as
+
+- API_KEY
+- AWS_IAM
+- AMAZON_COGNITO_USER_POOLS
+- OPENID_CONNECT.
+
+In this case, we are using an API key to protect the API, so we will need to create an API key.
+
+### Step 2: Create the Schema
+
+Now we have the API ready, we will need to create AppSync schema for the API
+
+```yml
+CoffeeShopAppSyncSchema:
+  Type: AWS::AppSync::GraphQLSchema
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    Definition: |
+      input orderInput {
+        id: String
+        coffee: String
+        size: String
+      }
+
+      type Order {
+        id: String
+        coffee: String
+        size: String
+      }
+
+      type Mutation {
+      createOrder(input: orderInput): Order
+      }
+
+      type Query{
+      getOrder(id: String): Order
+      }
+
+      schema {
+      query: Query
+      mutation: Mutation
+      }
+```
+
+The schema that we created serves the purpose of ensuring the data from the client matches the schema before mapping to the resolver.
+
+In the schema, we are creating a mutation for create order. This expects the input argument to match the type defined in orderInput. The result returned matches the type defined in Order.
+
+### Step 3: Create the Datasource
+
+The type of data sources supported by AppSync are
+
+- AMAZON_DYNAMODB,
+- AMAZON_ELASTICSEARCH
+- AWS_LAMBDA, HTTP
+- RELATIONAL_DATABASE
+
+In this case, we will use DynamoDb as our data source.
+
+```yml
+CoffeeShopDynamoDB:
+  Type: AWS::DynamoDB::Table
+  Properties:
+    TableName: "CoffeeShopTable"
+    AttributeDefinitions:
+      - AttributeName: "id"
+        AttributeType: "S"
+      - AttributeName: "resourceId"
+        AttributeType: "S"
+    KeySchema:
+      - AttributeName: "id"
+        KeyType: "HASH"
+      - AttributeName: "resourceId"
+        KeyType: "RANGE"
+    ProvisionedThroughput:
+      ReadCapacityUnits: 3
+      WriteCapacityUnits: 1
+    Projection:
+      ProjectionType: "ALL"
+
+CoffeeShopDynamoDBDataSource:
+  Type: AWS::AppSync::DataSource
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    Name: CoffeeShopDynamoDBDataSource
+    Description: Contain order data from coffee shop
+    Type: “AMAZON_DYNAMODB”
+    ServiceRoleArn: arn:aws:iam::xxxxxxxxxxxx:role/xxxxxxxxx
+    DynamoDBConfig:
+      TableName: CoffeeShopTable
+      AwsRegion: ap-southeash-2
+```
+
+From the template, we are provisioning new dynamoDB table for the API. It has the attribute id as the hash key and resourceId as the range key for the table.
+
+We are connecting the AppSync data source with the newly created dynamoDB table by specifying the table name. You can connect to an existing data source if you already have one.
```

Comment:
DynamoDB

Reply:
Also by saying "You can connect to an existing data source if you already have one.", it might be worth providing a link to setting up other data sources with AppSync but personal preference on this one.